folding-schemes/src/arith/r1cs/circuits.rs

```diff
@@ -0,0 +1,301 @@
+use crate::{
+    arith::ArithGadget,
+    utils::gadgets::{EquivalenceGadget, MatrixGadget, SparseMatrixVar, VectorGadget},
+};
+use ark_ff::PrimeField;
+use ark_r1cs_std::alloc::{AllocVar, AllocationMode};
+use ark_relations::r1cs::{Namespace, SynthesisError};
+use ark_std::{borrow::Borrow, marker::PhantomData, One};
+
+use super::R1CS;
+
+/// An in-circuit representation of the `R1CS` struct.
+///
+/// `M` is for the modulo operation involved in the satisfiability check when
+/// the underlying `FVar` is `NonNativeUintVar`.
+#[derive(Debug, Clone)]
+pub struct R1CSMatricesVar<M, FVar> {
+    _m: PhantomData<M>,
+    pub A: SparseMatrixVar<FVar>,
+    pub B: SparseMatrixVar<FVar>,
+    pub C: SparseMatrixVar<FVar>,
+}
+
+impl<F: PrimeField, ConstraintF: PrimeField, FVar: AllocVar<F, ConstraintF>>
```

Comment:
Great question! In fact, `FieldVar` would be overkill for `R1CSMatricesVar`, because `R1CSMatricesVar` only needs some simple operations on vectors and matrices defined in `{Vector, Matrix}Gadget`, but methods in `FieldVar` are never used.

Thus, when defining `R1CSMatricesVar`, `FVar` is used only as a placeholder, without any bounds like `FieldVar`. Later, when implementing methods for `R1CSMatricesVar`, only necessary trait bounds are added. For example:
- In the implementation of `AllocVar` for `R1CSMatricesVar`, `FVar` only needs to satisfy `AllocVar` due to the `new_constant` call, while methods defined in other traits (e.g., `VectorGadget`) are not used.
https://github.com/privacy-scaling-explorations/sonobe/blob/5c6e3c50ec61d1ce7d7771b6676506fe186e5dbc/folding-schemes/src/arith/r1cs/circuits.rs#L24-L43
- In the implementation of `eval_at_z`, `SparseMatrixVar<FVar>` should satisfy `MatrixGadget<FVar>`, and `[FVar]` should satisfy `VectorGadget<FVar>`, but `AllocVar` now becomes unnecessary.
https://github.com/privacy-scaling-explorations/sonobe/blob/5c6e3c50ec61d1ce7d7771b6676506fe186e5dbc/folding-schemes/src/arith/r1cs/circuits.rs#L45-L60

Such a design, as you mentioned, produces more concise code, because the caller only need to specify trait bounds that are necessary.

Another reason why I didn't use `FieldVar` is because of our customized `NonNativeUintVar`. Unlike `NonNativeFieldVar` or `EmulatedFpVar` that operate in a finite field, `NonNativeUintVar` just represents unsigned big integers, which do not have methods like `FieldVar::{negate, inverse}`. Therefore, `FVar` should not satisfy `FieldVar`, since otherwise, `SparseMatrixVar<NonNativeUintVar>` would become an invalid type for R1CS matrices.

Reply:
Oh nice! Thanks for the explanation!